backends/p4tools/testgen/core/small_step/table_stepper.h

```diff
@@ -0,0 +1,207 @@
+    /// This function returns false, if no key needs to be resolved.
+    /// If keys are tainted or a key needs to be resolved, this function returns true.
+    bool resolveTableKeys();
+
+    /// This function allows target back ends to add their own match types.
+    /// For example, some back ends implement the "optional" match type, which either hits as exact
+    /// match or does not match at all. The table stepper first checks these custom match types. If
+    /// these do not match it steps through the default implementation. If it does not match either,
+    /// a P4C_UNIMPLEMENTED is thrown.
+    virtual const IR::Expression* computeTargetMatchType(
+        ExecutionState* nextState, const KeyProperties& keyProperties,
+        std::map<cstring, const FieldMatch>* matches, const IR::Expression* hitCondition);
+
+    /// A helper function that computes whether a control-plane/table-key hits or not. This does not
+    /// handle constant entries, it is specialized for control plane entries.
+    /// The function also tracks the list of field matches created to achieve a  hit. We later use
+    /// this to insert table entries using the STF/PTF framework.
+    const IR::Expression* computeHit(ExecutionState* nextState, const IR::P4Table* table,
+                                     std::map<cstring, const FieldMatch>* matches);
+
+    /// Collects properties that may be set per table. Target back end may have different semantics
+    /// for table execution that need to be collect before evaluation the table.
+    virtual void checkTargetProperties(
+        const std::vector<const IR::ActionListElement*>& tableActionList);
+
+    /* =========================================================================================
+     * Table Evaluation functions
+     * ========================================================================================= */
+    /// When we hit a table.apply() method call expression we transition into the table call
+    /// phase. Handling table calls is fairly involved as we need to respect the control plane
+    /// manipulating these tables. This is why we are using several helper functions.
+    /// @param table the table we invoke.
+    void evalTableCall();
+
+    /// Tables may have constant match-action entries. This usually implies that the table is
+    /// immutable. This helper function unrolls all entries and aligns the table key with the match
+    /// provided by the constant entry.
+    /// @param table the table we invoke.
+    /// @returns tableMissCondition reference to the current constraints of the miss condition.
+    /// Any constant entry we hit implies that the table is hit, and the default action is not
+    /// executed.
+    const IR::Expression* evalTableConstEntries();
+
+    /// This helper function evaluates potential insertion from the control plane. We use zombie
+    /// variables to mimic an operator inserting entries. We only cover ONE entry per table for
```

Comment:
Does the comment with ONE belong here?
So in your tests a table has always at most one entry?

Reply:
Yup, covering exactly one table path. We may change this in the future but for now this is how the interpreter is designed.